backend/cfg/cfg_dataflow.ml

```diff
@@ -132,3 +132,192 @@ module Forward (D : Domain) (T : Transfer with type domain = D.t) :
+  SXXX with type domain = D.t = struct
+  (* CR xclerc for xclerc: see what can be shared with `Forward`. *)
+
+  type domain = D.t
+
+  type map = domain Label.Tbl.t
+
+  module WorkSetElement = struct
+    type t =
+      { label : Label.t;
+        value : domain
+      }
+
+    let compare { label = left_label; value = left_value }
+        { label = right_label; value = right_value } =
+      match Label.compare left_label right_label with
+      | 0 -> D.compare left_value right_value
+      | res -> res
+  end
+
+  module WorkSet = Set.Make (WorkSetElement)
+
+  let transfer_block : domain -> exn:domain -> Cfg.basic_block -> domain =
+   fun value ~exn block ->
+    ListLabels.fold_right block.body
+      ~init:(T.terminator value ~exn block.terminator) ~f:(fun instr value ->
+        T.basic value ~exn instr)
+
+  let is_exit_block : Cfg.basic_block -> bool =
+   fun block ->
+    match block.terminator.desc with
+    | Never -> assert false
+    | Always _ -> false
+    | Parity_test _ -> false
+    | Truth_test _ -> false
+    | Float_test _ -> false
+    | Int_test _ -> false
+    | Switch _ -> false
+    | Return -> true
+    | Raise _ -> block.trap_depth = 1
+    | Tailcall _ -> true
+    | Call_no_return _ -> true
+
+  let create : Cfg.t -> init:domain -> map * WorkSet.t ref =
+   fun cfg ~init ->
+    let map = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let set = ref WorkSet.empty in
+    let value = init in
+    Cfg.iter_blocks cfg ~f:(fun label block ->
+        if is_exit_block block || true
+           (* CR xclerc for xclerc: remove `|| true` *)
+        then begin
+          Label.Tbl.replace map label value;
+          set := WorkSet.add { WorkSetElement.label; value } !set
+        end);
+    map, set
+
+  let remove_and_return :
+      Cfg.t -> WorkSet.t ref -> WorkSetElement.t * Cfg.basic_block =
+   fun cfg set ->
+    let element = WorkSet.choose !set in
+    set := WorkSet.remove element !set;
+    element, Cfg.get_block_exn cfg element.label
+
+  let run :
+      Cfg.t -> ?max_iteration:int -> init:domain -> unit -> (map, map) Result.t
+      =
+   fun cfg ?(max_iteration = max_int) ~init () ->
+    let res, work_set = create cfg ~init in
+    let iteration = ref 0 in
+    (* note: `handler_map` contains the value at the *start* of the block. *)
+    let handler_map : D.t Label.Tbl.t =
+      Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks)
+    in
+    while (not (WorkSet.is_empty !work_set)) && !iteration < max_iteration do
+      incr iteration;
+      let element, block = remove_and_return cfg work_set in
+      let exn : domain =
+        Option.map
+          (fun exceptional_successor ->
+            Label.Tbl.find_opt handler_map exceptional_successor)
+          block.exn
+        |> Option.join
+        |> Option.value ~default:D.bot
+      in
+      let value = transfer_block element.value ~exn block in
+      if block.is_trap_handler
```

Comment:
Why is `trap_handler` block treated differently from a normal block? I expected the only difference to be in updating the `handler_map` for trap_handler blocks, but the propagation to be the same.